sae_lens/training/toy_models.py

```diff
@@ -271,17 +211,138 @@ def optimize(
 
             # Display progress bar
             if step % log_freq == 0 or (step + 1 == steps):
-                progress_bar.set_postfix(
-                    loss=loss.item() / self.cfg.n_instances, lr=step_lr
-                )
+                progress_bar.set_postfix(loss=loss.item(), lr=step_lr)
+
+
+class ReluOutputModel(HookedToyModel):
+    """
+    Anthropic's ReLU Output Model as described in the Toy Models paper:
+            https://transformer-circuits.pub/2022/toy_model/index.html#demonstrating-setup-model
+    """
+
+    W: Float[Tensor, "n_hidden n_features"]
+    b_final: Tensor
+    # Our linear map is x -> ReLU(W.T @ W @ x + b_final)
+
+    def __init__(self, cfg: ToyConfig, device: torch.device = torch.device("cpu")):
+        super().__init__(cfg)
+
+        self.W = nn.Parameter(
+            nn.init.xavier_normal_(t.empty((cfg.n_hidden, cfg.n_features)))
+        )
+        self.b_final = nn.Parameter(t.zeros((cfg.n_features)))
+        self.to(device)
+
+        # Add and setup hookpoints.
+        self.hook_hidden = HookPoint()
+        self.hook_out_prebias = HookPoint()
+        self.setup()
+
+    def forward(
+        self,
+        features: Float[Tensor, "... features"],
+        return_type: str | None = None,
+    ) -> Float[Tensor, "... features"]:
+        hidden = self.hook_hidden(
+            einops.einsum(
+                features,
+                self.W,
+                "... features, hidden features -> ... hidden",
+            )
+        )
+        out = self.hook_out_prebias(
+            einops.einsum(
+                hidden,
+                self.W,
+                "... hidden, hidden features -> ... features",
+            )
+        )
+        reconstructed = F.relu(out + self.b_final)
+
+        if return_type == "loss":
+            return self.calculate_loss(reconstructed, features)
+        else:
+            return reconstructed
+
+    def calculate_loss(
+        self,
+        out: Float[Tensor, "batch features"],
+        batch: Float[Tensor, "batch features"],
+    ) -> Float[Tensor, ""]:
+        """
+        Calculates the loss for a given batch, using this loss described in the Toy Models paper:
+
+            https://transformer-circuits.pub/2022/toy_model/index.html#demonstrating-setup-loss
+
+        Note, `model.importance` is guaranteed to broadcast with the shape of `out` and `batch`.
+        """
+        error = self.importance * ((batch - out) ** 2)
+        loss = einops.reduce(error, "batch features -> ()", "mean").sum()
+        return loss
```

Comment:
This loss calculation would be good to cover in a test, but this is optional. You're already increasing test coverage a lot with this PR. I'm also happy to add tests to this after merging.